packages/sources/cfbenchmarks-test/src/endpoint/websocket/crypto.ts

```diff
@@ -0,0 +1,150 @@
+import { makeLogger } from '@chainlink/external-adapter-framework/util'
+import {
+  WebSocketTransport,
+  WebSocketRawData,
+} from '@chainlink/external-adapter-framework/transports/websocket'
+import { WebSocket } from '@chainlink/external-adapter-framework/transports/websocket'
+import { EndpointTypes, getIdFromBaseQuote, getBaseQuoteFromId } from '../common/crypto'
+
+interface Message {
+  type: 'subscribe' | 'unsubscribe' | 'value'
+  id: string
+  value: string
+  time: number
+}
+export interface WsErrorType {
+  TYPE: string
+  MESSAGE: string
+  PARAMETER: string
+  INFO: string
+}
+
+export type WsEndpointTypes = EndpointTypes & {
+  Provider: {
+    WsMessage: Message
+  }
+}
+
+const logger = makeLogger('CfbenchmarksCryptoWebsocketEndpoint')
+
+export const transport = new WebSocketTransport<WsEndpointTypes>({
+  url: ({ adapterConfig: { DEFAULT_WS_API_ENDPOINT } }) => DEFAULT_WS_API_ENDPOINT,
+
+  options: ({ adapterConfig: { API_USERNAME, API_PASSWORD } }) => {
+    const encodedCreds = Buffer.from(`${API_USERNAME}:${API_PASSWORD}`).toString('base64')
+    return {
+      headers: {
+        Authorization: `Basic ${encodedCreds}`,
+      },
+    }
+  },
+
+  handlers: {
+    open(connection: WebSocket) {
+      return new Promise((resolve, reject) => {
+        // Set up listener
+        connection.on('message', (data: WebSocketRawData) => {
+          const parsed = JSON.parse(data.toString())
+          if (parsed.MESSAGE === 'STREAMERWELCOME') {
+            logger.info('Got logged in response, connection is ready')
+            resolve()
+          } else {
+            reject(new Error('Unexpected message after WS connection open'))
+          }
+        })
+      })
+    },
+
+    message(message) {
+      logger.trace(message, 'Got response from websocket')
+      if (message.type === 'value') {
+        const { base, quote } = getBaseQuoteFromId(message.id)
+        const value = Number(message.value)
+        return [
+          {
+            params: { base, quote },
```

Comment:
Just realized ... @alejoberardino does this cache based on the `params`? If so this may be a problem.
The WS message we get back is the id - we don't know whether it came from a request base/quote or id.
I guess we can cache on both and/or on every combo.

Reply:
Ah, good point. For these the way to make sure this is not an issue is to use the `Adapter.requestTransforms`: these will be applied to the parameters after validation but before cache keys are generated. What we should make sure is that if you would receive from the DP the same message for different params, we should be consolidating the params before in these transformations so that we can be sure it's a 1 to 1 relationship

Reply:
Not sure if I follow that second sentence.

Do you mean: when writing to the cache we need to match back up to the original params, but we consolidated them down, so we will need to mark `transport` and `type: primary | secondary` in the .requestTransforms and result params?

Reply:
Regardless, because of the primary vs secondary thing `requestTransform`s does not work cleanly for this adapter without changing the framework to also expose the `AdapterConfig`. I could read it directly from the environment, but definitely not  a good practice.

Reply:
I think exposing the config to the routing function makes sense, we should make the change to the framework on that one

Reply:
Made the change in the latest commit, check it out.

Reply:
Great, that should take care of the problem, now whether you request the EA with base+quote or id, after the request is validated and transformations applied, all it will use is the id, therefore making it 1-1 with the responses from the provider and always match requests to their responses in the cache